### PR TITLE
Add Language support

### PR DESCRIPTION
**Pull request by nguhoa**
Fix issue: https://github.com/microsoft/Power-Fx/issues/1128

### CODE REVIEW

**Comment by lesaltzm** [comment]
Very minorly Breaking Internals, requires cleaning up the corresponding definition in PA Client

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj [hunk 1]
```diff
@@ -48,6 +48,10 @@
     </Content>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj:51]: Remove this?

**Review by lesaltzm — DISMISSED** [review]

File: src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj (hunk 1 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj:51]: Remove this. thanks

**Comment by MikeStall** [comment]
+1 
Specifically, when we merge, we should be able to remove Microsoft.AppMagic.Authoring.Texl.LanguageFunction and use the definition added here. 


---
In reply to: [1456868937](https://github.com/microsoft/Power-Fx/pull/1147#issuecomment-1456868937) [](http://example.com/codeflow?ancestors=1456868937,1456868937)

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryLanguage.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    internal static partial class Library
+    {
+        public static FormulaValue Language(EvalVisitor runner, EvalVisitorContext context, IRContext irContext, FormulaValue[] args)
+        {
+            return Language(runner, irContext);
+        }
+
+        public static StringValue Language(EvalVisitor runner, IRContext irContext)
+        {
+            return new StringValue(irContext, runner.CultureInfo.ToString());
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryLanguage.cs:22]: >ToString [](http://example.com/codeflow?start=65&length=8)

Just double checking this is the correct string function. 
CultureInfo has several. .Name, .DisplayName, etc.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using Xunit;
+using static Microsoft.PowerFx.Functions.Library;
+
+namespace Microsoft.PowerFx.Tests
+{
+    public class LanguageTest : PowerFxTest
+    {
+        // Test language
+        [Fact]
+        public void GetLanguageTest()
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs:21]: >GetLanguageTest [](http://example.com/codeflow?start=20&length=15)

We should add "default" test for what happens when we don't set it. 

var engine = new RecalcEngine();
engine.Eval("Language()")

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryLanguage.cs (hunk 2 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryLanguage.cs:22]: update to .Name

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs (hunk 3 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs:21]: Add this test

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 4]
```diff
@@ -1593,6 +1593,17 @@ static Library()
...
+            },
+            {
+                BuiltinFunctionsCore.Language,
+                StandardErrorHandling<FormulaValue>(
+                    BuiltinFunctionsCore.Language.Name,
+                    expandArguments: NoArgExpansion,
+                    replaceBlankValues: DoNotReplaceBlank,
+                    checkRuntimeTypes: DeferRuntimeTypeChecking,
+                    checkRuntimeValues: DeferRuntimeValueChecking,
+                    returnBehavior: ReturnBehavior.ReturnEmptyStringIfAnyArgIsBlank,
```
Comment (anderson-joyle) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1605]: >ReturnEmptyStringIfAnyArgIsBlank [](http://example.com/codeflow?start=51&length=32)

Language function doesn't take any argument. Is this still the right option?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Language.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+>> Language()
+"en-US"
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Language.txt:2]: Can we get a different result if someone in France run this test?

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            runtimeConfig.SetTimeZone(TimeZoneInfo.Utc);
+            var runner = new EvalVisitor(runtimeConfig, CancellationToken.None);
+
+            var language = Language(runner, IRContext.NotInSource(FormulaType.String));
+            Assert.Equal(vnCulture.ToLower(), language.Value.ToLower());
+        }
+
+        // Test default language
+        [Fact]
+        public void GetDefaultLanguageTest()
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs:35]: >GetDefaultLanguageTest [](http://example.com/codeflow?start=20&length=22)

Missing Assert.*?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Microsoft.PowerFx.Core.Texl.Builtins
+{
+    internal class LanguageFunction : BuiltinFunction
+    {
+        public override bool IsSelfContained => true;
+
+        public override bool HasPreciseErrors => true;
+
+        public override bool SupportsParamCoercion => true;
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs:20]: >public override bool SupportsParamCoercion => true; [](http://example.com/codeflow?start=8&length=51)

Please remove it - it doesn't have any parameters

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Texl.Builtins
+{
+    internal class LanguageFunction : BuiltinFunction
+    {
+        public override bool IsSelfContained => true;
+
+        public override bool HasPreciseErrors => true;
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs:18]: >public override bool HasPreciseErrors => true; [](http://example.com/codeflow?start=8&length=46)

No need to override it: it doesn't have precise errors (nor has any parameters - all of its instances will be valid

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 4 above)
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1605]: You can use NoErrorHandling. Take a look at Pi, which has a similar signature.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs (hunk 8 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs:18]: This is just copying as is from PAClient.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs (hunk 7 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs:20]: same as above

File: src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs [hunk 9]
```diff
@@ -216,6 +216,7 @@ internal class BuiltinFunctionsCore
         public static readonly TexlFunction WeekNum = _library.Add(new WeekNumFunction());
         public static readonly TexlFunction With = _library.Add(new WithFunction());
         public static readonly TexlFunction Year = _library.Add(new YearFunction());
+        public static readonly TexlFunction Language = _library.Add(new LanguageFunction());
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs:219]: >public static readonly TexlFunction Language = _library.Add(new LanguageFunction()); [](http://example.com/codeflow?start=8&length=84)

Please move to line 115 (to maintain the list ordered)

Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs:219]: updated this.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 4 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1605]: updated this. thanks

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 10]
```diff
@@ -1593,6 +1593,10 @@ static Library()
                     checkRuntimeValues: DeferRuntimeValueChecking,
                     returnBehavior: ReturnBehavior.AlwaysEvaluateAndReturnResult,
                     targetFunction: Year)
+            },
+            {
+                BuiltinFunctionsCore.Language,
+                NoErrorHandling(Pi)
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1599]: >Pi [](http://example.com/codeflow?start=32&length=2)

`Language`

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Language.txt (hunk 5 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Language.txt:2]: No, we have default cultrure set for test is en-Us

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs (hunk 6 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs:35]: we should still assert the return value.  (I should have clarified in my earlier comment)

var engine = new RecalcEngine();
var result = engine.Eval("Language()")
Assert.Equals("?????', result.ToObject());

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 10 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: updated.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs (hunk 6 above)
Comment (nguhoa) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageTest.cs:35]: Add assert. thanks

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs (hunk 8 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs:18]: removed.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs (hunk 7 above)
Comment (nguhoa) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Language.cs:20]: removed.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Review by MikeStall — APPROVED** [review]

**Pull request merged**